templates/fine-tune-function-calling/fc_utils/eval_utils.py

```diff
@@ -0,0 +1,163 @@
+
+from fc_utils.response_parsers import (
+    ERROR_OUTPUT,
+    AnyscaleResponseParser,
+    OpenAIResponseParser,
+    INCORRECT_FORMAT
+)
+
+
+class PossibleMistakes(Enum):
+    UNWANTED_FUNCTION_CALL = "Unwanted Function Call"
+    NO_FUNCTION_CALL = "No Function Call"
+    INCORRECT_FORMAT = "Incorrect Function Format"
+    INCORRECT_NUMBER_OF_FUNCTION_CALLS = "Incorrect Number of Function Calls"
+    WRONG_FUNCTION_NAME = "Wrong Function Name"
+    WRONG_ARGUMENT_VALUE = "Wrong Argument Value"
+    MISSING_ARGUMENT = "Missing Argument"
+
+    def values():
+        return [item.value for item in PossibleMistakes]
+
+def is_match(response: dict, ground_truth: dict) -> Tuple[bool, str]:
+    """
+    Checks if the response matches the ground truth. Returns a boolean and a string indicating the mistake type if any.
+    """
+    if ground_truth["tool_calls"] is None:
+        if response["tool_calls"] is None:
+            return True, ""
+        else:  # explicit else for clarity
+            return False, PossibleMistakes.UNWANTED_FUNCTION_CALL.value
+
+    if response["tool_calls"] is None:
+        return False, PossibleMistakes.NO_FUNCTION_CALL.value
+    elif response["tool_calls"] == INCORRECT_FORMAT: # error during parsing
+        return False, PossibleMistakes.INCORRECT_FORMAT.value
+    elif len(response["tool_calls"]) != len(ground_truth["tool_calls"]):
+        return False, PossibleMistakes.INCORRECT_NUMBER_OF_FUNCTION_CALLS.value
+
+    for expected_tool_call, actual_tool_call in zip(ground_truth["tool_calls"], response["tool_calls"]):
+        if "name" not in actual_tool_call or "arguments" not in actual_tool_call:
+            return False, PossibleMistakes.INCORRECT_FORMAT.value
+        if expected_tool_call["name"] != actual_tool_call["name"]:
+            return False, PossibleMistakes.WRONG_FUNCTION_NAME.value
+        elif expected_tool_call["arguments"] != actual_tool_call["arguments"]:
+            if len(expected_tool_call["arguments"]) != len(actual_tool_call["arguments"]):
+                return False, PossibleMistakes.MISSING_ARGUMENT.value
+            return False, PossibleMistakes.WRONG_ARGUMENT_VALUE.value
+    return True, ""
+
+
+def parse_and_eval(
+    parser: Union[OpenAIResponseParser, AnyscaleResponseParser],
+    messages,
+    expected_responses,
+    tools=None,
+):
+    """
+    Parse and eval loop to parse the assistant responses and evaluate them against the ground truth in the conversation.
+    Args:
+    parser: OpenAIResponseParser or AnyscaleResponseParser object
+    messages: list of messages in the conversation. The messages with role 'assistant' are ignored and triggers a chat completion call
+    expected_responses: list of ground truth responses
+    tools: list of tools to available, used by OpenAIResponseParser
+    """
+    assert messages[0]["role"] == "system", "First message must be from system"
+    match = True
+    current_conv = []
+    assist_id = 0
+    conv_id = 0
```

Comment:
nit: for index use idx or index not id. also name them: `cur_message_idx` and `cur_expected_idx`